### DISCUSSION TITLE
Get an error run `scripts/test` in a newly cloned `uvicorn`

### DISCUSSION THREAD

**Discussion by ys-wu** [Kludex/uvicorn#1755] [Potential Issue] [answered]
I clone my fork: 
```
git clone https://github.com/MY-USERNAME/uvicorn
```

Then simply install and test:
```
cd uvicorn
scripts/install
scripts/test
```

And get the error:
```
+ '[' -z ']'
+ scripts/check
+ ./scripts/sync-version
+ venv/bin/black --check --diff --target-version=py37 uvicorn tests
All done! ✨ 🍰 ✨
63 files would be left unchanged.
+ venv/bin/flake8 uvicorn tests
+ venv/bin/mypy --show-error-codes
Success: no issues found in 53 source files
+ venv/bin/isort --check --diff --project=uvicorn uvicorn tests
+ venv/bin/python -m tools.cli_usage --check
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/Users/wuyusheng/playground/uvicorn/tools/cli_usage.py", line 69, in <module>
    rv |= _generate_cli_usage(path, check=args.check)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/wuyusheng/playground/uvicorn/tools/cli_usage.py", line 47, in _generate_cli_usage
    usage_lines = _get_usage_lines()
                  ^^^^^^^^^^^^^^^^^^
  File "/Users/wuyusheng/playground/uvicorn/tools/cli_usage.py", line 13, in _get_usage_lines
    res = subprocess.run(["uvicorn", "--help"], stdout=subprocess.PIPE)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/wuyusheng/.pyenv/versions/3.11.0/lib/python3.11/subprocess.py", line 546, in run
    with Popen(*popenargs, **kwargs) as process:
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/wuyusheng/.pyenv/versions/3.11.0/lib/python3.11/subprocess.py", line 1022, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/Users/wuyusheng/.pyenv/versions/3.11.0/lib/python3.11/subprocess.py", line 1899, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: 'uvicorn'
```

My environment:
```
MacBook Pro 2017
macOS Monterey Version 12.6 (21G115)
Python 3.11.0
```

**Comment by ys-wu** [comment]
However, this problem doesn't exist if I activate a virtual environment: 
`activate venv/bin/activate`
and run:
`python -m tools.cli_usage --check`

**Reply by ys-wu** [reply]
I found the answer, and the virtual environment must be activated first.

**Reply by Kludex** [reply]
You shouldn't be having this issue... :thinking:

**Reply by ys-wu** [reply]
Thank you for your reply, but I don't understand what you mean.
To me, the error is reproducible.

**Comment by iudeen** [comment]
Got curious and got a fresh Linux machine (Ubuntu 22.04 with Python 3.10) to verify this.

I was able to reproduce this issue there. Entire log dump here FYR: https://pastebin.ubuntu.com/p/cpBNZhHF9J/

**Reply by ys-wu** [reply]
Yes, it is reproducible in 2 different machines of mine.
If the error is not reproducible for some of you, maybe you had `uvicorn` installed already before cloned.

My understanding:
Line 13 in `tools/cli_usage.py`:
```python
res = subprocess.run(["uvicorn", "--help"], stdout=subprocess.PIPE)`
```
is not using `uvicorn` from `venv`.
So It gives the error if there is no `uvicorn` outside `venv`.

My solution:
Instead of using:
```python
 res = subprocess.run(["uvicorn", "--help"], stdout=subprocess.PIPE)
```
Modify it to the following works for me:
```python
res = subprocess.run([sys.executable.replace("python", "uvicorn"), "--help"], stdout=subprocess.PIPE)
```

Or, simply, activate the virtual environment before running `scripts/test`.

Please correct me if I'm wrong.

**Reply by Kludex** [reply]
@ys-wu If you apply those modifications, and you run `./scripts/lint`, do you have any changes in the documentation? If yes, then it's not enough. :cry: 

If not... PR welcome.

**Comment by Kludex** [comment, marked as answer]
This is an issue. See more on: https://github.com/encode/uvicorn/issues/1763#issuecomment-1328602872.
